src/temporal/interval/index.ts

```diff
@@ -0,0 +1,86 @@
+import { Temporal } from "temporal-polyfill";
+
+declare module "temporal-polyfill" {
+  namespace Temporal {
+    class Interval<T extends Temporal.PlainDate | Temporal.ZonedDateTime> {
+      static from<T extends Temporal.PlainDate | Temporal.ZonedDateTime>(start: T, end: T): Interval<T>;
+
+      readonly start: T;
+      readonly end: T;
+
+      constructor(start: T, end: T);
+
+      eachMonth(options?: { step: number }): T[];
+      eachWeek(options?: { step?: number; weekStartsOn?: number }): T[];
+      contains(value: T, options?: { excludeBoundaries?: boolean }): boolean;
+    }
+  }
+}
+
+Temporal.Interval = class<T extends Temporal.PlainDate | Temporal.ZonedDateTime> {
```

Comment:
Would it be valid to ask for a `Interval<DateTime>`? That seems fine except such an interval could not have an `eachMonth` method?

This makes me think that this should be like `CalendarInterval` to denote it's for calendar-aware types, and keep `Interval` available for future / eventual `Interval<DateTime>` use case?

Reply:
why couldn't it have an `eachMonth` method? I could see zoned date times eventually wanting something like `.eachHour` / `.eachSecond` / `.eachMinute`, but we could restrict those with `this: ...` typeguards.

Do you mean `PlainTime`?

Reply:
Sorry, I was thinking of `Interval<Instant>`.

> we could restrict those with this: ... typeguards

True, but seems much more boring-in-a-good-way to just have `ZonedDateTimeInterval`, and in general a more traditional `Interval` (only `contains`)  -> `CalendarInterval` -> `ZonedDateTimeInterval` type hierarchy that progressively adds in more specific methods.

I really enjoyed using Java library back in the day:

https://timeandmoney.sourceforge.net/

Ah yeah, still have a fork of it around:

https://github.com/stephenh/timeandmoney/blob/master/src/main/java/com/domainlanguage/intervals/Interval.java

https://github.com/stephenh/timeandmoney/blob/master/src/main/java/com/domainlanguage/time/CalendarInterval.java

Reply:
Not sure why you'd want to work with `Instant` from business logic.  `Instant` is basically a time since epoch timestamp.  The moment you start using it you open yourself up to making all sorts of mistakes that zoned/plain types protect you from.

Reply:
Seems like kind of a deflection; `Instant` is a type in temporal, and `Interval<Instant>` seems like not an inconceivable ask.

I feel it's pretty objective that `Interval`-no-calendar-awareness -> `CalendarInterval`-calendar-awareness is an improvement/more correct than putting calendar-awareness on `Interval` and just saying "it should never be used for anything else".

Reply:
I don't really think it is a deflection. The entire point of `Temporal` is to get away from the everything is a timestamp approach of JS Dates.  I don't think we should be building support for `Instant` beyond interop with the built in dates.

Reply:
I did go back and forth on whether `Interval` should have 1:1 sub types for each of the base `Temporal` types, or at least for `ZonedDateTime` and `PlainDate`, but ended up deciding against it just so to keep things DRY.   If you really feel strongly that there should be more `Interval` types I can give it a shot, but I don't think `Instant` is one that should be supported.  Its use should be discouraged.

Reply:
> The entire point of Temporal is to get away from the everything is a
> timestamp approach of JS Dates

Agreed, the purpose is getting away from "(plaindate | zoneddatetime | instant) are all the same types at runtime" (and in the type system).

But I don't think that extends to "...and also instant should never be used".

I don't feel strongly about having `Interval<Instant>` officially supported in this PR, but I do feel strongly that we should be able to add `Interval<Instance>` support at some point in the future, and not be a breaking API change because we committed the `Interval` name/API contract to an "effectively calendared type".

Does renaming `Interval` to `CalendarInterval`, and indefinitely tabling/deferring `Interval<Instance>` sound okay?

Reply:
> But I don't think that extends to "...and also instant should never be used".

I disagree. There is no benefit to using them as anything more than a transitory type.  There are only downsides and footguns. I feel like you might just not have really looked deeply into the API?  You really can't do very much with `Instant` other than convert them into other types.  The few methods that do exist are very explicit they only work on units of hours, minutes, seconds, or smaller.  All the same methods and properties are also available on `ZonedDateTime`, so there is really zero reason to ever use them.  In the thousands of changes I made to graphql-service, I found zero uses for them.  

> Does renaming `Interval` to `CalendarInterval`, and indefinitely tabling/deferring Interval<Instance> sound okay?

I just still don't really see the benefit? It just seems like a longer name. Even if I was going to implement per type Intervals, I would want to nest them under the owning type's namespace, eg `Temporal.PlainDate.Interval`

Reply:
Future proofing for what? The spec has taken years to get to stage 3, a bunch of new types or dramatic changes to existing ones aren't going to suddenly emerge.  Again, I feel like you are basing this on your experience with prior APIs and not on how `Temporal` is actually constructed (eg, see [object model](https://tc39.es/proposal-temporal/docs/object-model.svg)).  The only two types that aren't "calendered" are `PlainTime` and `Instant`. I've yet to find a use for either (beyond `Date` conversion), and have already expressed why I think use of `Instant` should not be encouraged.  Maybe `PlainTime` is actually useful (some sort of scheduling?), but I don't think we need to go out of the way to future proof for it (ie, we could just have `ClockInterval` or something if we really had to).  Additionally, `Temporal` already has interfaces (ie, `Duration`) that are shared between every date/time type and the parts of those interfaces each type can't use are simply ignored/not provided.

Reply:
Future proofing for a `Interval<Instant>`.

And, again, clarity: seeing `eachMonth` on a seems-generic/not-obviously-calendared type, in a library (temporal) specifically dedicated to strictly delineating calendared/non-calendared types, would, for me, violate least-surprise.

I.e. it would make my brain slow down and go "what? why does a non-calendared type support this? ... oh right, it is actually calendared".

I anticipate eventually making a breaking change to fix this, but :shrug: I'll approve.

Reply:
> Future proofing for a `Interval<Instant>`

Why? Can you give any concrete example where that would actually be uniquely useful (ie, couldn't just be done with `ZonedDateTime`) and not a footgun?

> I.e. it would make my brain slow down and go "what? why does a non-calendared type support this? ... oh right, it is actually calendared".

Did your brain slow down and go "what?" when you saw that `Duration` has `years` / `months` / `days` /`weeks` properties then? Should it be renamed to `CalendarDuration`? It seems idiomatic of the spec to have a single interface shared by all types that simply isn't fully implemented for all of them.  This is still javascript

> I anticipate eventually making a breaking change to fix this, but 🤷 I'll approve.

How much money would you like to bet and what is the time frame? 😃